proxy/coco/src/announce.rs

```diff
@@ -0,0 +1,303 @@
+//! Compute, track and announce noteworthy changes to the network.
+
+use std::collections::HashSet;
+use std::ops::Deref as _;
+
+use kv::Codec as _;
+
+use librad::keys;
+use librad::net;
+use librad::uri;
+
+use crate::error::Error;
+use crate::oid::Oid;
+use crate::peer;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (uri::RadUrn, String, Oid);
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(
+    protocol: net::protocol::Protocol<net::peer::PeerStorage<keys::SecretKey>, net::peer::Gossip>,
+    updates: Vec<Announcement>,
+) -> Result<(), Error> {
+    for (project_urn, head, hash) in &updates {
+        let urn = uri::RadUrn::new(
+            project_urn.id.clone(),
+            uri::Protocol::Git,
+            uri::Path::parse(head)?,
+        );
+        let have = net::peer::Gossip {
+            urn,
+            rev: Some(net::peer::Rev::Git((*hash).into())),
+            origin: Some(protocol.peer_id().clone()),
```

Comment:
This should be `None`.

Sorry for this being confusing -- this field is actually meant to send a gossip message saying "I just fetched stuff from X (but I'm not X), go get it from me if you want". Iow, if I'm `Y`, and I'm interested in `X` (but not `Z`), but `Z` tells me they got some `X` stuff, I'll go look for `remotes/X/*` on `Z`. Or rather, I'll only bother fetching from `Z` if the update is about `X`. At least that's what's _supposed_ to happen.

Reply:
It does make sense, guess it could be better named and/or different types for the different message constructions. As in have one where we craft a gossip msg for another peer and one where we build one for the current peer.

Reply:
I'm open to suggestions, but note that the difference is not other vs current peer: if you set `origin`, you're saying that _you_ have `remotes/<origin>/*`.